dockers/texcmp/texcmp.js

```diff
@@ -0,0 +1,244 @@
+"use strict";
+
+var querystring = require("querystring");
+var child_process = require("child_process");
+var fs = require("fs");
+var path = require("path");
+var Q = require("q"); // To debug, pass Q_DEBUG=1 in the environment
+var pngparse = require("pngparse");
+var fft = require("ndarray-fft");
+var ndarray = require("ndarray-fft/node_modules/ndarray");
+
+var data = require("../../test/screenshotter/ss_data");
+
+// ignore some tests, since they contain commands not supported by LaTeX
+var blacklist = {
+    Colors: "Color handling differs",
+    DeepFontSizing: "\\Huge inside \\dfrac doesn't work for some reason",
+    KaTeX: "Custom command, doesn't exist in LaTeX"
+};
+var todo;
+if (process.argv.length > 2) {
+    todo = process.argv.slice(2);
+} else {
+    todo = Object.keys(data).filter(function(key) {
+        return !blacklist[key];
+    });
+}
+
+// Dimensions used when we do the FFT-based alignment computation
+var alignWidth = 2048; // should be at least twice the width resp. height
+var alignHeight = 2048; // of the screenshots, and a power of two.
+
+// Compute required resolution to match test.html. 16px default font,
+// scaled to 4em in test.html, and to 1.21em in katex.css. Corresponding
+// LaTeX font size is 10pt. There are 72.27pt per inch.
+var pxPerEm = 16 * 4 * 1.21;
+var pxPerPt = pxPerEm / 10;
+var dpi = pxPerPt * 72.27;
+
+var tmpDir = "/tmp/texcmp";
+var ssDir = path.normalize(
+    path.join(__dirname, "..", "..", "test", "screenshotter"));
+var imagesDir = path.join(ssDir, "images");
+var teximgDir = path.join(ssDir, "tex");
+var diffDir = path.join(ssDir, "diff");
+var template;
+
+Q.all([
+    Q.nfcall(fs.readFile, path.join(ssDir, "test.tex"), "utf-8"),
+    mkdir(tmpDir),
+    mkdir(teximgDir),
+    mkdir(diffDir)
+]).spread(function(data) {
+    template = data;
+    // dirs have been created, template has been read, now rasterize.
+    return Q.all(todo.map(processTestCase));
```

Comment:
At some point in the past, `template` was the result of a call to `readFileSync`, so this is partly historic. But we might get back to something like this one day, for some reason I can#t imagine right now, and then the global variable is more flexible because we don't have to care where it gets set, as long as it does get set by the time processTestCase is invoked.